kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.mwam.kafkakewl.metrics.services
+
+import com.mwam.kafkakewl.metrics.MainConfig
+import com.mwam.kafkakewl.metrics.domain.{KafkaTopicPartition, KafkaTopicPartitionInfo, KafkaTopicPartitionInfoChanges}
+import zio.kafka.consumer.Consumer
+import zio.{Duration, Fiber, Schedule, Task, UIO, URIO, ZIO, ZLayer}
+import zio.stream.{ZSink, ZStream}
+import zio.metrics.*
+import zio.metrics.Metric.Counter
+
+import scala.collection.immutable.Map
+
+object Helper {
+  def newTopicMetric(topic: String): Metric.Counter[Long] = Metric.counter("kafkakewl_vnext_topic_high_offset").tagged("topic", topic)
+  def newPartitionMetric(topic: String, partition: Int): Metric.Counter[Long] =
+    Metric.counter("kafkakewl_vnext_topic_partition_high_offset").tagged("topic", topic).tagged("partition", partition.toString)
+}
+
+case class MetricMap(metricMap: Map[String, (Map[Int, (Metric.Counter[Long], Long)], Metric.Counter[Long])]) {
+
+  def updateWithTopicParitionChanges(
+      changes: KafkaTopicPartitionInfoChanges,
+      config: MetricExposerConfig
+  ): (MetricMap, List[(Metric.Counter[Long], Long)]) = {
+    val addedOrUpdated = changes.addedOrUpdated
+    val removed = changes.removed
+    var newMetricMap = metricMap
+    var metricUpdates: List[(Metric.Counter[Long], Long)] = List.empty
```

Comment:
It's a functional list, appending is expensive as far as I remember. But I guess order does not matter really, so we can just prepend below instead of appending? Alternatively you can use a mutable list something like `ListBuffer` and make it immutable at the end.

Reply:
Fixed to be prepending